### PR TITLE
Configurable metrics reporter by catalog properties

### PR DESCRIPTION
**Pull request by kmozaid**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseTable.java
Comment (gaborkaszab) [inline core/src/main/java/org/apache/iceberg/BaseTable.java:66]: What I see in RESTSessionCatalog is that the MetricsReporter itself is not exposed from the class. Rather a MetricsReport is exposed through a function that is given to the BaseTable during creation:
https://github.com/apache/iceberg/blob/6697129a314d98cb793601495f7ebb2ae000b40a/core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java#L315

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (gaborkaszab) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:215]: Is it necessary to create the table through a transaction? Isn't this the result of copy-pasting from the thest above that tests transactional create table? :)

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (nastra) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:218]: rather than exposing `reporter()` in `BaseTable`, I think it would be better to actually make sure that the custom reporter's `report()` method is being called

File: core/src/main/java/org/apache/iceberg/BaseTable.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseTable.java:66]: I don't think we should expose this. We should rather test that `report()` of the custom reporter is actually being called

File: api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java
Comment (kmozaid) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java:27]: sure, I can rename, As of now I don't have use case to supply any properties, though one could pass prometheus gateway url or any other rest api where custom metrics reporter can push. (although these info can be directly read for static properties file that custom metric reporter will be dependent on.

File: core/src/main/java/org/apache/iceberg/BaseTable.java
Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseTable.java:66]: sure.

Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseTable.java:66]: removing this function.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (kmozaid) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:215]: I don't think, its necessary to use transaction. Let me change it.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (kmozaid) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:218]: sure.

**Comment by nastra** [comment]
@kmozaid now that #6404 is merged, can you please rebase this PR?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (kmozaid) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:218]: I don't see see a way to assert that custom reporter's `report()` method is being called if we don't expose this field. Do we use reflection to get field hidden fields and assert on them?

File: api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java
Comment (kmozaid) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsReporter.java]: for now, I have removed method.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java
Comment (kmozaid) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java]: Since, I don't see a way to assert, I have removed this test.

**Comment by kmozaid** [comment]
> @kmozaid now that #6404 is merged, can you please rebase this PR?

Hi @nastra I have rebased and updated PR with feedbacks.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:56]: I don't think we would want to init a reporter on every single `loadTable()` call. We should rather have a lazy `reporter` field that is initialized once. Something like this will work:

```
diff --git a/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java b/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
index 16a5164f6..8d5258620 100644
--- a/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
+++ b/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
@@ -25,6 +25,8 @@ import org.apache.iceberg.exceptions.AlreadyExistsException;
 import org.apache.iceberg.exceptions.CommitFailedException;
 import org.apache.iceberg.exceptions.NoSuchTableException;
 import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.metrics.LoggingMetricsReporter;
+import org.apache.iceberg.metrics.MetricsReporter;
 import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
@@ -35,6 +37,19 @@ import org.slf4j.LoggerFactory;
 
 public abstract class BaseMetastoreCatalog implements Catalog {
   private static final Logger LOG = LoggerFactory.getLogger(BaseMetastoreCatalog.class);
+  private MetricsReporter reporter;
+
+  private MetricsReporter reporter() {
+    if (null == reporter) {
+      this.reporter =
+          properties().containsKey(CatalogProperties.METRICS_REPORTER_IMPL)
+              ? CatalogUtil.loadMetricsReporter(
+                  properties().get(CatalogProperties.METRICS_REPORTER_IMPL))
+              : LoggingMetricsReporter.instance();
+    }
+
+    return reporter;
+  }
 
   @Override
   public Table loadTable(TableIdentifier identifier) {
@@ -51,7 +66,7 @@ public abstract class BaseMetastoreCatalog implements Catalog {
         }
 
       } else {
-        result = new BaseTable(ops, fullTableName(name(), identifier));
+        result = new BaseTable(ops, fullTableName(name(), identifier), reporter());
       }
 
     } else if (isValidMetadataIdentifier(identifier)) {
```

**Review by nastra — CHANGES_REQUESTED** [review]
For testing the custom metrics reporter it would be good to add the following test to `TestJdbcCatalog`:

```
+  @Test
+  public void testCatalogWithCustomMetricsReporter() throws IOException {
+    catalog =
+        initCatalog(
+            "test_jdbc_catalog_with_custom_reporter",
+            ImmutableMap.of(
+                CatalogProperties.METRICS_REPORTER_IMPL, CustomMetricsReporter.class.getName()));
+
+    catalog.buildTable(TABLE, SCHEMA).create();
+    Table table = catalog.loadTable(TABLE);
+    table
+        .newFastAppend()
+        .appendFile(
+            DataFiles.builder(PartitionSpec.unpartitioned())
+                .withPath("/path/to/data-a.parquet")
+                .withFileSizeInBytes(10)
+                .withRecordCount(2)
+                .build())
+        .commit();
+
+    try (CloseableIterable<FileScanTask> tasks = table.newScan().planFiles()) {
+      assertThat(tasks.iterator()).hasNext();
+    }
+
+    // counter of custom metrics reporter should have been increased
+    // 1x for commit metrics / 1x for scan metrics
+    assertThat(CustomMetricsReporter.COUNTER.get()).isEqualTo(2);
+  }
+
+  public static class CustomMetricsReporter implements MetricsReporter {
+    static final AtomicInteger COUNTER = new AtomicInteger(0);
+
+    @Override
+    public void report(MetricsReport report) {
+      COUNTER.incrementAndGet();
+    }
+  }
```

**Comment by nastra** [comment]
@kmozaid while reviewing this PR I've noticed that not every catalog actually exposes its properties (including the JDBC catalog which I suggested above to use for testing). I've opened https://github.com/apache/iceberg/pull/6471 to address that

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java]: thanks

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:317]: nit: newline missing after the closing if stmt

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:309]: nit: I don't think this needs to be synchronized

**Review by nastra — APPROVED** [review]
LGTM with some nits. 
@rdblue could you review this one please?

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java]: ok

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:318]: added newline.

**Comment by kmozaid** [comment]
Can I please get another review for merge to get enabled?

**Review by PraveenNanda124 — COMMENTED** [review]
Looks good

**Comment by kmozaid** [comment]
@gaborkaszab Can you please get this merged?

**Comment by gaborkaszab** [comment]
> @gaborkaszab Can you please get this merged?

The patch LGTM, however, I won't be able to merge as I'm not a committer. cc @rdblue

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:315]: Looks like this will enable LoggingMetricsReporter for all catalogs now, no way to turn it off.  @kmozaid @nastra curious is it the desired behavior?

Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:315]: HI @szehon-ho Even without this change, LoggingMetricsReporter is always enabled for all catalogs...See BaseTable https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/BaseTable.java#L45

Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:315]: Makes sense, sorry I didnt have the context but wonder is there enough value to always fill the logs every time?  But agree its not part of this issue.

**Comment by szehon-ho** [comment]
I think its fine with me if we can fix the failures

**Comment by kmozaid** [comment]
@gaborkaszab @szehon-ho  Could you please trigger workflows?

**Comment by nastra** [comment]
@kmozaid could you rebase the PR please? 
@danielcweeks could you review this please when you get a chance (and also trigger CI)?

**Comment by kmozaid** [comment]
@nastra I have rebased PR.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:468]: I'm confused, because the changes in this file should be gone after a rebase

Comment (kmozaid) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java]: sorry, I didn't build locally.
thanks, I agree it should be gone. Fixed it.

**Comment by szehon-ho** [comment]
Merged, thanks @kmozaid , also @nastra , @gaborkaszab , @PraveenNanda124 for reviews

**Pull request merged**
